icepyx/core/visualization.py

```diff
@@ -0,0 +1,503 @@
+
+            for fname in filelist:
+
+                if self.product in ["ATL07", "ATL10"]:
+                    format_string = (
+                        "{product:5d}-{HH:02d}_{datetime:%Y%m%d%H%M%S}_{rgt:04d}{cycle:02d}{"
+                        "region:02d}_{release:03d}_{version:02d}.h5"
+                    )
+                else:
+                    format_string = (
+                        "{product:5d}_{datetime:%Y%m%d%H%M%S}_{rgt:04d}{cycle:02d}{"
+                        "region:02d}_{release:03d}_{version:02d}.h5"
+                    )
+                temp = intake.source.utils.reverse_format(
+                    format_string=format_string, resolved_string=fname
+                )
+
+                rgt = temp["rgt"]  # Reference Ground Track
+                cycle = temp["cycle"]  # Cycle number
+                f_date = str(temp["datetime"].date())  # Datetime
+
+                paras = [rgt, f_date, cycle, bbox_i, self.product]
+                paras_list.append(paras)
+
+        return paras_list
+
+    @backoff.on_exception(
+        backoff.expo,
+        (
+            requests.exceptions.Timeout,
+            requests.exceptions.ConnectionError,
+            requests.exceptions.ChunkedEncodingError,
+        ),
+        max_tries=5,
+    )
+    def make_request(self, base_url, payload):
+        """
+        Make HTTP request
+        
+        Parameters
+        ----------
+        base_url : string
+            OpenAltimetry URL
+
+        See Also
+        --------
+        request_OA_data
+        """
+        return requests.get(base_url, params=payload)
+
+    def request_OA_data(self, paras) -> da.array:
+        """
+        Request data from OpenAltimetry based on API:
+        https://openaltimetry.org/data/swagger-ui/#/
+
+        Parameters
+        ----------
+        paras : list
+            A single parameter list for an OpenAltimetry API data request.
+
+        Returns
+        -------
+        OA_darr : da.array
+            A dask array containing the ICESat-2 elevation data.
+        """
+
+        base_url = "https://openaltimetry.org/data/api/icesat2/level3a"
+        trackId, Date, cycle, bbox, product = paras
+
+        # Generate API
+        payload = {
+            "product": product.lower(),
+            "endDate": Date,
+            "minx": str(bbox[0]),
+            "miny": str(bbox[1]),
+            "maxx": str(bbox[2]),
+            "maxy": str(bbox[3]),
+            "trackId": trackId,
+            "outputFormat": "json",
+        }  # default return all six beams
+
+        # request OpenAltimetry
+        r = self.make_request(base_url, payload)
+
+        # get elevation data
+        elevation_data = r.json()
+
+        df = pd.json_normalize(data=elevation_data, record_path=["data"])
+
+        # get data we need (with the correct date)
+        try:
+
+            df_series = df.query(expr="date == @Date").iloc[0]
+            beam_data = df_series.beams
+
+        except:
```

Comment:
Are there multiple exceptions we're capturing here, or can we explicitly list the one we're targeting?

Reply:
Yes, sorry I forgot to specify them, they are `NameError` and `KeyError` caused by empty dataframes return from `pd.json_normalize(data=elevation_data, record_path=["data"])`, could you please add them in your end? I am afraid it may cause chaos if I push changes from my side atm......thanks!

```
try:

    df_series = df.query(expr="date == @Date").iloc[0]
    beam_data = df_series.beams

except (NameError, KeyError) as error:
    beam_data = None
```

Reply:
It should be safe to push the `except (NameError, KeyError)` change @icetianli (as long as you do a local `git pull` from the `openaltimetry` branch first). Jessica's recent changes didn't modify L347 so the `git revert` commands I mentioned at https://github.com/icesat2py/icepyx/pull/144#issuecomment-829686206 won't cause any problems.

Reply:
OK, done